### PR TITLE
Changes to bring testing numbers up and other stuff

### LINKED ISSUE #45: Collaborator Mutation - verify uniqueness of collaboration
Ensure a collaborator isn't already added.

### PR DESCRIPTION
**Pull request by zachjamesgreen**
Change type for collaborators on project type
add project repo type to project type
Fixing small inaccuracies and other small bugs
other small changes

## Checklist
- [x] Tests written where required
- [x] SimpleCov at 85.35%
- [x] Squash & Rebase
- [x] All tests passing
- [ ] Prettier has been run
- [ ] Reek has been run
- [x] Circle CI builds are passing

Closes #45

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -19,12 +19,12 @@ jobs:
     steps:
       - checkout
       - restore_cache:
-          key: rails-demo-{{ checksum "Gemfile.lock" }}
+          key: codeherd-{{ checksum "Gemfile.lock" }}
```
Comment (twarbelow) [inline .circleci/config.yml:22]: is this fixing the deploy issue?

File: app/graphql/types/project_type.rb [hunk 2]
```diff
@@ -5,7 +5,8 @@ class ProjectType < Types::BaseObject
     field :summary, String, null: true
     field :mod_number, String, null: true
     field :owner, Types::UserType, null: true
-    field :collaborators, [Types::UserType], null: true
+    field :collaborators, [Types::CollaboratorType], null: true
```
Comment (twarbelow) [inline app/graphql/types/project_type.rb:8]: thanks for catching this

File: app/graphql/types/query_type.rb [hunk 3]
```diff
@@ -17,15 +17,15 @@ class QueryType < Types::BaseObject
...
-    field :user, Types::UserType, null: false do
-      description "return a user"
-      argument :id, ID, required: true
-    end
-
     def users
       User.all
     end
 
+    field :user, Types::UserType, null: false do
```
Comment (twarbelow) [inline app/graphql/types/query_type.rb:24]: what prompted moving this? would love to know more.

File: app/graphql/types/query_type.rb [hunk 4]
```diff
@@ -34,16 +34,16 @@ def user(id:)
 
     field :users_projects, [Types::ProjectType], null: false do
       description "all of a user's projects"
-      argument :user_id, ID, required: true
+      argument :user_id, ID, required: true, description: "ID of the user you want to find projects for"
```
Comment (twarbelow) [inline app/graphql/types/query_type.rb:37]: NICE

File: app/graphql/types/query_type.rb [hunk 5]
```diff
@@ -34,16 +34,16 @@ def user(id:)
...
 
     def users_projects(user_id:)
       Project.where(owner_id: user_id)
     end
 
     field :project, Types::ProjectType, null: false do
-      description "the name of the project"
-      argument :id, ID, required: true
+      description "Get project by ID"
+      argument :id, ID, required: true, description: "ID of the project you want to find"
```
Comment (twarbelow) [inline app/graphql/types/query_type.rb:46]: SO GOOD

File: app/graphql/types/user_type.rb [hunk 6]
```diff
@@ -8,6 +8,7 @@ class UserType < Types::BaseObject
     field :working_styles, [String], null: true
     field :cohort, String, null: true
     field :pronouns, String, null: true
+    field :image, String, null: true
```
Comment (twarbelow) [inline app/graphql/types/user_type.rb:11]: are we saving the user image now?

File: app/models/collaborator.rb [hunk 7]
```diff
@@ -2,6 +2,6 @@ class Collaborator < ApplicationRecord
   belongs_to :user
   belongs_to :project
 
-  validates :user_id, presence: true#, uniqueness: true
-  validates :project_id, presence: true#, uniqueness: true
+  validates :user_id, presence: true, uniqueness: { scope: :project_id }
```
Comment (twarbelow) [inline app/models/collaborator.rb:5]: would you be able to explain a little more about how this is working? I understand the end result, but could use clarification...
- why does the scoped uniqueness need to be on both user_id and project_id? would having it on just one still prevent duplicates?

File: app/models/user.rb [hunk 8]
```diff
@@ -3,6 +3,9 @@ class User < ApplicationRecord
   validates :name, presence: true
   validates :github_handle, presence: true, uniqueness: true
   validates :github_access_token, presence: true, uniqueness: true
+  # validates :slack_handle, presence: true
+  # validates :cohort, presence: true
+  has_many :projects, dependent: :destroy, foreign_key: :owner_id
```
Comment (twarbelow) [inline app/models/user.rb:8]: excellent catch with the dependent destroy

File: public/collaborator.doc.html [hunk 9]
```diff
@@ -284,7 +284,7 @@ <h2 id="require-by" class="graphdoc-section__title slds-text-heading--medium sld
             </a>
             Require by
         </h2>
-        <ul class="require-by"><li><a href="createcollaborationpayload.doc.html" title="CreateCollaborationPayload - Autogenerated return type of CreateCollaboration">CreateCollaborationPayload<em>Autogenerated return type of CreateCollaboration</em></a><li></ul>
+        <ul class="require-by"><li><a href="createcollaborationpayload.doc.html" title="CreateCollaborationPayload - Autogenerated return type of CreateCollaboration">CreateCollaborationPayload<em>Autogenerated return type of CreateCollaboration</em></a><li><li><a href="project.doc.html" title="Project - ">Project<em>null</em></a><li></ul>
```
Comment (twarbelow) [inline public/collaborator.doc.html:287]: could you walk me through (on zoom maybe) how you made this documentation and how to update it/work with it

File: spec/models/collaborator_spec.rb [hunk 10]
```diff
@@ -8,8 +8,8 @@
 
   describe 'validations' do
     it {should validate_presence_of(:user_id)}
-    # it {should validate_uniqueness_of(:user_id)}
     it {should validate_presence_of(:project_id)}
-    # it {should validate_uniqueness_of(:project_id)}
+    # it {should validate_uniqueness_of(:user_id).scoped_to(:project_id)}
```
Comment (twarbelow) [inline spec/models/collaborator_spec.rb:12]: why don't we  test this?

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -7,6 +7,9 @@
   it {should validate_presence_of(:github_access_token)}
   it {should validate_uniqueness_of(:github_handle)}
   it {should validate_uniqueness_of(:github_access_token)}
+  # it {should validate_presence_of(:slack_handle)}
```
Comment (twarbelow) [inline spec/models/user_spec.rb:10]: do we need to leave these commented out lines in here?

File: spec/rails_helper.rb [hunk 12]
```diff
@@ -66,6 +66,17 @@
   # config.filter_gems_from_backtrace("gem name")
 
   config.include FactoryBot::Syntax::Methods
+
```
Comment (twarbelow) [inline spec/rails_helper.rb:69]: is there a reason these database cleaner lines are commented out?

File: spec/request/github_auth_spec.rb [hunk 13]
```diff
@@ -7,8 +7,7 @@
 
   it 'should create a user' do
     count = User.count
-    get '/sessions/new'
-    follow_redirect!
+    get '/auth/github'
     follow_redirect!
```
Comment (twarbelow) [inline spec/request/github_auth_spec.rb:11]: I've never seen `follow_redirect!`...interesting

File: spec/request/mutations/create_collaboration_spec.rb [hunk 14]
```diff
@@ -54,6 +54,19 @@ def query(user_id, email, project_id)
     expect(parsed[:data][:createCollaboration][:errors][0]).to eq("Validation failed: User must exist")
   end
 
+  it 'cannot create a collaborator if collaboration already exists' do
```
Comment (twarbelow) [inline spec/request/mutations/create_collaboration_spec.rb:57]: thank you for dealing with this, I appreciate it

File: spec/request/mutations/edit_user_spec.rb [hunk 15]
```diff
@@ -39,4 +39,27 @@ def query(id)
     json = JSON.parse(response.body)
     expect(json['data']['editUser']['errors']).to include("Couldn't find User with 'id'=0")
   end
+
+  xit 'returns error if user cannot be updated' do
```
Comment (twarbelow) [inline spec/request/mutations/edit_user_spec.rb:43]: if we are not using this test, should we remove it?

**Review by twarbelow — COMMENTED** [review]
Hey Zach,
Left some questions....

File: .circleci/config.yml (hunk 1 above)
Comment (zachjamesgreen) [reply .circleci/config.yml:22]: no, its me trying to figure out how cacheing works in CircleCI. Also is there a deploy issue that I'm not aware of?

File: app/graphql/types/query_type.rb (hunk 3 above)
Comment (zachjamesgreen) [reply app/graphql/types/query_type.rb:24]: IDK just wanted to make it match all the other code written in this file

File: app/graphql/types/user_type.rb (hunk 6 above)
Comment (zachjamesgreen) [reply app/graphql/types/user_type.rb:11]: yes

File: app/models/collaborator.rb (hunk 7 above)
Comment (zachjamesgreen) [reply app/models/collaborator.rb:5]: Yeah it just makes sure that a user_id project_id pair is unique. I could just have it on one. Just covering all the bases

File: spec/models/user_spec.rb (hunk 11 above)
Comment (zachjamesgreen) [reply spec/models/user_spec.rb:10]: might put it back in later

File: spec/rails_helper.rb (hunk 12 above)
Comment (zachjamesgreen) [reply spec/rails_helper.rb:69]: Its not working the way I want it to

File: spec/request/mutations/edit_user_spec.rb (hunk 15 above)
Comment (zachjamesgreen) [reply spec/request/mutations/edit_user_spec.rb:43]: There is one line of code not tested that this test will cover just need to work out how to have a user fail to update

File: .circleci/config.yml (hunk 1 above)
Comment (twarbelow) [reply .circleci/config.yml:22]: I thought you had said that it wasn't set up for automatic deploy.

File: app/models/collaborator.rb (hunk 7 above)
Comment (twarbelow) [reply app/models/collaborator.rb:5]: cool, thanks for the explanation

**Pull request merged**